### ISSUE TITLE
Maps with non-literal keys cannot be serialised into JsonObject

### ISSUE THREAD

**Issue by antohaby** [Kotlin/kotlinx.serialization#542]
**Describe the bug**
When attempting to serialise an object that has some Map with non-literal key to JsonObject causes following Exception:

```
Exception in thread "main" java.lang.IllegalStateException: Expected JsonLiteral, but has {"id":"42"}
	at kotlinx.serialization.json.internal.JsonTreeMapOutput.putElement(TreeJsonOutput.kt:151)
	at kotlinx.serialization.json.internal.AbstractJsonTreeOutput$beginStructure$consumer$1.invoke(TreeJsonOutput.kt:92)
	at kotlinx.serialization.json.internal.AbstractJsonTreeOutput$beginStructure$consumer$1.invoke(TreeJsonOutput.kt:21)
	at kotlinx.serialization.json.internal.AbstractJsonTreeOutput.endEncode(TreeJsonOutput.kt:109)
	at kotlinx.serialization.TaggedEncoder.endStructure(Tagged.kt:103)
	at com.example.SomeId$$serializer.serialize(Application.kt)
	at com.example.SomeId$$serializer.serialize(Application.kt:8)
....
```

Stringify works well. 

**To Reproduce**

```kotlin
@Serializable
data class SomeId(
    val id: String
)

@Serializable
data class Container(
    val map: Map<SomeId,String>
)

fun main(args: Array<String>) {
    val container = Container(
        mapOf(
            SomeId("42") to "24"
        )
    )
    val json = Json(JsonConfiguration.Stable)
    val jsonString = json.stringify(Container.serializer(), container) // works well
    println(jsonString) // {"map":[{"id":"42"},"24"]}
    val jsonObject = json.toJson(Container.serializer(), container) // fails
}
```

**Expected behavior**
Expected to have JsonObject with map serialized as an array of (key,value) tuples. (same like in stringify).

**Environment**
 - Kotlin version: 1.3.41
 - Library version: 0.11.1
 - Kotlin platforms: JVM (didn't test on other platforms)
 - Gradle version: 5.4.1

**Comment by antohaby** [comment]
Cause:

```kotlin
// class TreeJsonOutput
        val encoder = when (desc.kind) {
            StructureKind.LIST, UnionKind.POLYMORPHIC -> JsonTreeListOutput(json, consumer)
            // Here is the problem. 
           // in case if typeParams[0].descriptor.kind is not a primitive kind
           // JsonTreeListOutput has to be return instead JsonTreeMapOutput
            StructureKind.MAP -> JsonTreeMapOutput(json, consumer)
            else -> JsonTreeOutput(json, consumer)
        }
```

In fact I would suggest to reuse JsonStreamingLogic (which responsible for json strings) as much as possible. or at least try to make it generic.

**Comment by pdvrieze** [comment]
For non-literal keys the map would need to work differently, it couldn't actually serialize as a json map, but would have to do so as a list of key value pairs.

Btw. in this particular case, it may be better to give your ID type a custom serializer that works as if it is a primitive (the id value only).

**Comment by antohaby** [comment]
I'm fine with list of pairs.  Of course I can have custom serialiser with `StringDescriptor` or any other workarounds to avoid this issue. 

But that is actually a bit broader. I also got another exception when I tried to do polymorphic serialization with discriminator field. So the overall `TreeJsonOutput` is broken because it is out of sync with `JsonStreamingLogic` (which is stable).

**Comment by sandwwraith** [comment]
It is probably fixed in today's `0.12.0` release with the `JsonConfiguration` flag `allowStructuredMapKeys`
